### PR TITLE
versions: Switch to virtio-fs-dev branch for kernel

### LINKED ISSUE #2603: Upgrade kernel used for virtiofs
The upgrade should pull in newer kernel(5.6) and necessary changes for overlay support for virtiofs.

### PR DESCRIPTION
**Pull request by amshinde**
This includes a newer kernel and necessary overlay changes
that fix oustanding issue for running docker in docker using
overlay.

Fixes #2603

Signed-off-by: Archana Shinde <archana.m.shinde@intel.com>

### CODE REVIEW

**Comment by amshinde** [comment]
Added 
```
 Depends-on: github.com/kata-containers/agent#738
 Depends-on: github.com/kata-containers/runtime#2463
```
to take into account the runtime and agent changes required for this PR.

**Comment by amshinde** [comment]
@fidencio Looks like CI is still failing. Should https://github.com/kata-containers/runtime/pull/2473 be included as well for CI to pass?

**Comment by fidencio** [comment]
> @fidencio Looks like CI is still failing. Should #2473 be included as well for CI to pass?

It should and I was pretty sure I did that Yesterday. Let me re-do it again, by editing your comment.

**Comment by fidencio** [comment]
Added
```
 Depends-on: github.com/kata-containers/runtime#2473
```

**Comment by fidencio** [comment]
@amshinde, I have the feeling the issues are stull happening as we do need the patched agent vendored in the shim. What's the best way to proceed with this?

**Comment by fidencio** [comment]
Added
```
 Depends-on: github.com/kata-containers/shim#233
```

**Comment by fidencio** [comment]
There's something fishy happening here. I can see in the logs that either editing your comment or add the deps in different comments, the PR is only considering as deps:
```
13:51:52 This PR depends on repository: github.com/kata-containers/agent and pull request: 738
...
13:51:53 This PR depends on repository: github.com/kata-containers/runtime and pull request: 2463
```

In reality, we should have 4 PRs as dependency, considering the one touching the vendored code.

**Comment by amshinde** [comment]
@fidencio The dependency needs to be part of the commit message. I'll add it there and retrigger CI.

**Comment by fidencio** [comment]
Hmm. I see a different error now:
```
Stderr: rpc error: code = Internal desc = Could not resolve symlink for source kataShared
```

@amshinde, do we ensure, somehow, that the image / initrd being used was generated using the patched agent (with the patch applied in the vendored part)? How do we do that?

**Comment by amshinde** [comment]
@fidencio Yes, it should generate a new rootfs. I do see the rootfs generation in the logs. 
The new agent should get copied into the rootfs. @chavafg Can you confirm.
But I am seeing this in the logs:
```
INFO: current kernel : virtio-fs-v0.3-72-virtiofs
09:01:04 INFO: cached kernel  : virtio-fs-v0.3-72-virtiofs
09:01:04 INFO: Install pre-built kernel version
09:01:04 INFO: Installing vmlinux
```

@chavafg Is it not using the new kernel for virtio-fs ?

@fidencio Have you tested the combination of PRs locally to make sure we not missing anything?

**Comment by chavafg** [comment]
about the patched agent, yes we are using it... you can see that we are applying PR 738 and rebasing with master
```
This PR depends on repository: github.com/kata-containers/agent and pull request: 738
/tmp/jenkins/workspace/kata-containers-runtime-ubuntu-18-04-virtiofs-PR/go/src/github.com/kata-containers/agent /tmp/jenkins/workspace/kata-containers-runtime-ubuntu-18-04-virtiofs-PR/go/src/github.com/kata-containers/runtime /tmp/jenkins/workspace/kata-containers-runtime-ubuntu-18-04-virtiofs-PR/go/src/github.com/kata-containers/runtime
Fetching pull request: 738 for repository: github.com/kata-containers/agent
From https://github.com/kata-containers/agent
 * [new ref]         refs/pull/738/head -> p738
Switched to branch 'p738'
First, rewinding head to replay your work on top of it...
Applying: mount: Use "virtiofs" instead of "virtio_fs"
baef95d mount: Use "virtiofs" instead of "virtio_fs"
73afd1a Merge pull request #740 from justin-he/strip
cfc96dd build: Add support to strip the agent golang binary
```
and then we can see that the image to be generated is:
`INFO: Image to generate: kata-containers-clearlinux-32820-osbuilder-891b61c-agent-baef95d.img
`, which corresponds to the commit ID of the agent: `baef95d mount: Use "virtiofs" instead of "virtio_fs"`.

The kernel installation seems to be taking the cached version instead of the one specified in this PR. 
Checking this furtherly.

**Comment by fidencio** [comment]
@amshinde, that's exactly what we have on Fedora nowadays:
- agent: https://src.fedoraproject.org/rpms/kata-agent/blob/master/f/kata-agent.spec#_56
  - patch: https://src.fedoraproject.org/rpms/kata-agent/blob/master/f/0001-mount-Use-virtiofs-instead-of-virtio_fs-as-typeVirti.patch
- shim: https://src.fedoraproject.org/rpms/kata-shim/blob/master/f/kata-shim.spec#_70
  - patch: https://src.fedoraproject.org/rpms/kata-shim/blob/master/f/0001-mount-Use-virtiofs-instead-of-virtio_fs-as-typeVirti.patch
- runtime: https://src.fedoraproject.org/rpms/kata-runtime/blob/master/f/kata-runtime.spec#_78
  - patches: https://src.fedoraproject.org/rpms/kata-runtime/blob/master/f/0002-kata_agent-Use-virtiofs-instead-of-virtio_fsas-typeV.patch and https://src.fedoraproject.org/rpms/kata-runtime/blob/master/f/0003-kata_agent-Drop-sharedDirVirtioFS-Options.patch

Second patch from Fedora is different, as the version submitted does not add "dax" to the sharedDirVirtioFSDaxOptions, but in Fedora "dax" is **never** appended (as QEMU in Fedora doesn't know about "dax" option at all). Also, the error for this would be different if it was related.

**Comment by chavafg** [comment]
@fidencio @amshinde, I see that you added a `Depends-on` to this repo PR 2473, so it is taking that PR and discarding changes in this one. Right now our CI does not properly works testing changes of 2 PRs of the same repository. 
So that is why the CI is running the virtiofs job with the old (cached) kernel...
What about having #2473 changes in this PR and close that one?

File: versions.yaml [hunk 1]
```diff
@@ -158,9 +158,9 @@ assets:
     version: "v5.4.15"
 
   kernel-experimental:
-    description: "Linux kernel with virtiofs 3.0"
+    description: "Linux kernel with virtiofs-dev"
     url: "https://gitlab.com/virtio-fs/linux.git"
-    tag: "virtio-fs-v0.3"
+    tag: "virtio-fs-dev"
```
Comment (chavafg) [inline versions.yaml:163]: btw, could we use a commit ID here, instead of the branch name? So that we can get reproducible builds over time.

**Comment by fidencio** [comment]
> @fidencio @amshinde, I see that you added a `Depends-on` to this repo PR 2473, so it is taking that PR and discarding changes in this one. Right now our CI does not properly works testing changes of 2 PRs of the same repository.
> So that is why the CI is running the virtiofs job with the old (cached) kernel...
> What about having #2473 changes in this PR and close that one?

I'm more than fine. IMHO, we could have all my patches for runtime as part of this PR.
@amshinde, do you take care of this?

File: versions.yaml (hunk 1 above)
Comment (amshinde) [reply versions.yaml]: @chavafg Good suggestion. I have included a commit here instead. ptal.

**Comment by amshinde** [comment]
@fidencio I have included your commits here.
Lets see how this goes.

**Comment by fidencio** [comment]
```
22:53:52 + info 'Install experimental kernel'
22:53:52 + echo -e 'INFO: Install experimental kernel'
22:53:52 INFO: Install experimental kernel
22:53:52 + pushd /tmp/install-kata-RDVmxOozEcy
22:53:52 + /tmp/jenkins/workspace/kata-containers-runtime-ubuntu-18-04-virtiofs-PR/go/src/github.com/kata-containers/packaging/kernel/build-kernel.sh -e setup
22:53:52 /tmp/jenkins/workspace/kata-containers-runtime-ubuntu-18-04-virtiofs-PR/go/src/github.com/kata-containers/tests /tmp/install-kata-RDVmxOozEcy
22:53:52 /tmp/install-kata-RDVmxOozEcy
22:53:52 INFO: Config version: 72
22:53:52 INFO: Kernel version: 
22:53:52 INFO: kernel path does not exist, will download kernel
22:53:52 ERROR: failed to get kernel version: Kernel version is emtpy
22:53:52 ++ handle_error 72
22:53:52 ++ local exit_code=1
22:53:52 ++ local line_number=72
22:53:52 ++ echo 'Failed at 72: "${kernel_repo_dir}/kernel/build-kernel.sh" -e setup'
22:53:52 Failed at 72: "${kernel_repo_dir}/kernel/build-kernel.sh" -e setup
22:53:52 ++ exit 1
22:53:52 + exit_handler
22:53:52 + rm -rf /tmp/install-kata-RDVmxOozEcy
22:53:52 Failed at 30: "${cidir}/install_kata_kernel.sh" "${tag}"
22:53:52 Build step 'Execute shell' marked build as failure
```

**Comment by fidencio** [comment]
- **jenkins-ci-ARM-ubuntu-18-04**:
```
01:45:43 [reset] Deleting contents of stateful directories: [/var/lib/etcd /var/lib/kubelet /var/lib/dockershim /var/run/kubernetes /var/lib/cni]
01:45:43 
01:45:43 The reset process does not clean CNI configuration. To do so, you must remove /etc/cni/net.d
01:45:43 
01:45:43 The reset process does not reset or clean up iptables rules or IPVS tables.
01:45:43 If you wish to reset iptables, you must do so manually by using the "iptables" command.
01:45:43 
01:45:43 If your cluster was setup to utilize IPVS, run ipvsadm --clear (or similar)
01:45:43 to reset your system's IPVS tables.
01:45:43 
01:45:43 The reset process does not clean your kubeconfig files and you must remove them manually.
01:45:43 Please, check the contents of the $HOME/.kube/config file.
01:45:44 Bad argument `[unsupported'
01:45:44 Error occurred at line: 54
01:45:44 Try `iptables-restore -h' or 'iptables-restore --help' for more information.
01:45:44 Failed at 19: sudo iptables-restore < "$iptables_cache"
01:45:44 Failed at 39: bash -f "${SCRIPT_PATH}/cleanup_bare_metal_env.sh"
01:45:44 Failed at 1: popd
01:45:44 Makefile:148: recipe for target 'kubernetes' failed
01:45:44 make: *** [kubernetes] Error 2
01:45:44 Failed at 82: sudo -E PATH="$PATH" bash -c "make test"
01:45:44 Build step 'Execute shell' marked build as failure
01:45:44 Performing Post build task...
```
The same erros can be seen on the ARM tests run for https://github.com/kata-containers/runtime/pull/2606, what makes me quite confident to **not** be related to this PR.

**Comment by amshinde** [comment]
@fidencio Yes ARM CI seems to be failing consistently, but the error seems unrelated to this PR.
Lets merge this once we get another ack on this, followed by the agent PR and a revendor PR in the shim.
cc @chavafg

File: versions.yaml (hunk 1 above)
Comment (amshinde) [reply versions.yaml:163]: @chavafg Looks like `install_kernel` script in packaging repo expects a tag for the experimental kernel. I have reverted back to using a branch. Lets use a commit id in a future PR after we fix the script. Wdyt?

Comment (chavafg) [reply versions.yaml:163]: ok, sure

**Comment by amshinde** [comment]
@chavafg Can we merge this PR?

**Comment by chavafg** [comment]
sure, lets merge this one.

**Pull request merged**